### ISSUE TITLE
 Unable to cast object of type 'System.Security.Cryptography.RSACng' to type 'System.Security.Cryptography.RSACryptoServiceProvider'

### ISSUE THREAD

**Issue by catsaremine** [jwt-dotnet/jwt#222]
I have a cert that worked with 4.0.0 but now fails with 5.2.3.

```
InvalidCastException: Unable to cast object of type 'System.Security.Cryptography.RSACng' to type 'System.Security.Cryptography.RSACryptoServiceProvider'.
  Module "JWT.Algorithms.RS256Algorithm", line 0, col 0, in GetPublicKey
    System.Security.Cryptography.RSACryptoServiceProvider GetPublicKey(System.Security.Cryptography.X509Certificates.X509Certificate2)
  Module "JWT.Algorithms.RS256Algorithm", line 0, col 0, in .ctor
    Void .ctor(System.Security.Cryptography.X509Certificates.X509Certificate2)
  File "xxxxxxxxxxx\src\WebTest\Auth\JwtService.cs", line 46, col 13, in EncodeSession
    System.String EncodeSession(T)

```

**Comment by abatishchev** [comment]
Per [this question on SO](https://stackoverflow.com/questions/55949510/casting-private-key-to-rsacryptoserviceprovider-not-working) the object you're passing might not have the key set exportable. Please check.

**Comment by abatishchev** [comment]
By the way what platform you're on, .NET Framework or .NET Core. and of which version?

**Comment by abatishchev** [comment]
Trying to figure this out and fix in #223

**Comment by abatishchev** [comment]
Can you please submit a PR with a test that would help to reproduce the issue you're facing?

The problem in the PR I opened is that RsaCryptoProvider has overlords of VerifyData() which aren't present on RSA. So I need more time/help to figure this out.

Your input in both directions would be much appreciated.

**Comment by abatishchev** [comment]
Here's [the code](https://github.com/jwt-dotnet/jwt/blob/15ed092634fe3c21e680c8d4df371992d5f7b865/src/JWT/Algorithms/RS256Algorithm.cs) of RSAlgorithm that corresponds to version 4.0.0.0. I see the cast to `RSACryptoServiceProvider` is already there. Please help me to figure out what has been changed and what broke your code.

**Comment by abatishchev** [comment]
Here's [the version](https://github.com/jwt-dotnet/jwt/blob/c7a7145dfe8314dae1dbf8d18c5e87e9ae273357/src/JWT/Algorithms/RS256Algorithm.cs) of RS256Algorithm with the cast to `RSA` and not `RSACryptoServiceProvider`. I'm trying to recall what was the rationale behind this change.

**Comment by abatishchev** [comment]
Please try version [5.3.0-beta1](https://www.nuget.org/packages/JWT/5.3.0-beta1) (would need to enable to show pre-realeases).

I still would like you to submit a repro test to make sure this scenario is covered going forward.

**Comment by catsaremine** [comment]
I'm using netcoreapp2.2.

Everything works on 4.0.0, the certificate is the same.

Using  5.3.0-beta1, it encodes successfully but throws on verification: 

InvalidCastException: Unable to cast object of type 'System.Security.Cryptography.RSACng' to type 'System.Security.Cryptography.RSACryptoServiceProvider'.
  File "D:\a\1\s\src\JWT\Algorithms\RSAlgorithmFactory.cs", line 26, col 21, in Create
    JWT.Algorithms.IJwtAlgorithm Create(JWT.JwtHashAlgorithm)
  File "D:\a\1\s\src\JWT\JwtDecoder.cs", line 242, col 13, in Validate
    Void Validate(JWT.JwtParts, Byte[])
  File "D:\a\1\s\src\JWT\JwtDecoder.cs", line 90, col 17, in Decode
    System.String Decode(System.String, Byte[], Boolean)
  File "D:\a\1\s\src\JWT\JwtDecoder.cs", line 187, col 13, in DecodeToObject
    T DecodeToObject[T](System.String, Byte[], Boolean)
  File "xxxxxxxxxxxxxx\Auth\JwtService.cs", line 35, col 17, in DecodeSession
    T DecodeSession(System.String)

**Comment by abatishchev** [comment]
Oh, I see. I forgot to update the factory. I'll change it today. Can you please meanwhile try to instantiate the algorithm directly?

**Comment by abatishchev** [comment]
Please post the code how you're creating the certificate, I'll put it into a test.

**Comment by catsaremine** [comment]
I'm reading it from a file. 

```

        private static X509Certificate2 GetCertificate()
        {
            return new X509Certificate2(Path.Combine(Global.DataDirectory, "certificate.pfx"));
        }

```

**Comment by abatishchev** [comment]
Please try version [5.3.0-beta2](https://www.nuget.org/packages/JWT/5.3.0-beta2).

**Comment by catsaremine** [comment]
It works correctly with 5.3.0-beta2. Both encoding and verification.

**Comment by abatishchev** [comment]
Sound great! Please help a little and submit a PR to make it non-beta.

**Comment by abatishchev** [comment]
@catsaremine ping. As as cutesy to the open source community, and after I fixed the issue you were facing as soon as I could, please submit a trivial PR to share the results with the rest of the lib's consumers.

**Comment by catsaremine** [comment]
I will try to generate a certificate that breaks on stable and runs on beta.

I was testing with our real certificates and that is why I didn't write any tests the last N times you asked me.

Perhaps I should have said so. Sorry for the misunderstanding, I don't want to look like I'm ungrateful, because I am grateful. 

I will try to find how the certs were generated.

**Comment by abatishchev** [comment]
Resolved by #226.
